OS spec (#1053)
inngest/inngest: docs/SDK_SPEC.md

They are used to express relative time periods to an Inngest Server that doesn’t rely on the time of an SDK’s platform being accurate.

### 1.3.3. Composite ID

A Composite ID is a combination of a Function’s ID and the ID of the App that contains it, joined together with a `"-"`.

For example, if an App’s ID is `"myapp"` and a Function’s ID is `"myfn"`, the Composite ID of the Function would be `"myapp-myfn"`.

Composite IDs are used to reference Functions across App boundaries.

### 1.3.4. Event

An Event is a standardized JSON payload that represents a single event within an Inngest Server. It can be used to trigger one or more functions and can be sent from server- or client-side code.

An Event always has the following structure:

```tsx
{
  /**
   * A unique ID used to idempotently process a given event payload.
   *
   * A developer can set this when sending events to ensure that the event is
   * only processed once; if an event with the same ID is sent again, it will
   * not trigger any functions.
   */
	id: string;

  /**

>    * A unique IDfor the type of event. We recommend using lowercase dot

Reviewer: ```suggestion
   * A unique ID for the type of event. We recommend using lowercase dot
```

2. If a Signing Key has not been provided by a Developer, the SDK MUST reject the request and return a `500 Internal Server Error` response.
3. If the SDK is intending to use Inngest Cloud [4.2] and no `X-Inngest-Signature` header has been given, the SDK MUST reject the request and return a `500 Internal Server Error` response.
4. The SDK SHOULD then extract the `t` timestamp from the `X-Inngest-Signature` header and ensure it is a time within the last 5 minutes according to the SDK’s known time.

If this is checked and the timestamp is not within this time period, the SDK MUST reject the request and return a `500 Internal Server Error` response.
5. The SDK then MUST calculate an HMAC with SHA256 by specifying the key as the `s` value of the `X-Inngest-Signature` header with the `signkey-*-` prefix removed and hashing the body of the request followed by the timestamp `t` with hex encoding.

Note that if the raw bytes of the request body are inaccessible, the body should first be parsed using the JSON Canonicalization Scheme (JCS) as specified in [RFC 8785](https://datatracker.ietf.org/doc/html/rfc8785).

> If the generated hash does not exactly match the `s` value in the `X-Inngest-Signature` header, the SDK MUST reject the requets with a `500 Internal Server Error`.

Reviewer: ```suggestion
If the generated hash does not exactly match the `s` value in the `X-Inngest-Signature` header, the SDK MUST reject the requests with a `500 Internal Server Error`.
```

If these steps have been adhered to, the request is now deemed valid and can be used to trigger the flows that follow.

### 4.1.4. Requirements when sending a request

When sending any request to Inngest, an SDK MUST adhere to the requirements when responding to requests [4.1.2].

The SDK also MUST send a bearer token authorization request header, as defined in [RFC 6750](https://datatracker.ietf.org/doc/html/rfc6750#section-2.1). The value of this bearer token should be the Signing Key, where the value following the `signkey-*-` prefix is a hex-encoded SHA256 hash of that value.

For example:

```bash
# original signing key
signkey-prod-8fjau3mn

# hashed signing key
signkey-prod-3c8335d113497a3a0b3e6bc18c12bf59e3db1c964c9f66765f374f5f7b473ac7

# header
Authorization: Bearer signkey-prod-3c8335d113497a3a0b3e6bc18c12bf59e3db1c964c9f66765f374f5f7b473ac7
```

## 4.2. Kinds of Inngest Server

Inngest provides both a hosted Cloud (production) service and a Dev Server which allows local development of Inngest Functions. We refer to both as an Inngest Server.

> ### 4.2.1. Targetting an Inngest Server

Reviewer: ```suggestion
### 4.2.1. Targeting an Inngest Server
```

If this default is followed, the SDK MUST use the `INNGEST_DEV` [3.2] environment variable to allow the user to connect to a Dev Server, which can optionally also include a runtime configuration option.

The `INNGEST_DEV` environment variable may have a non-empty value, including an origin as defined in [RFC 6454](https://datatracker.ietf.org/doc/html/rfc6454), for example `http://example.com`.  If the `INNGEST_DEV` value is a valid host, the SDK should use this origin as the Dev Server URL.  Otherwise, for a non-zero `INNGEST_DEV` variable the SDK MUST attempt to connect to the following default URLs:

- API Origin: `http://localhost:8288`
- Event ingestion endpoint: `http://localhost:8288/e/:event_key`

If the `INNGEST_DEV` value is a vcalid [RFC 6454](https://datatracker.ietf.org/doc/html/rfc6454) origin, for example `http://example.com`, the resulting endpoints will then be used:

- API Origin: `http://example.com`
- Event ingestion endpoint: `http://example.com/e/:event_key`

### 4.2.2. Request verification

When receiving requests expected to be from an Inngest Server, an SDK MUST verify their integrity [4.1.3]. An Inngest Dev Server, however, MUST NOT require verification.

> If the recommendations when targetting an Inngest Server [4.2.1] are followed, then presence of the `INNGEST_DEV` environment variable (or equivalent runtime configuration option) MUST turn off signature verification.

Reviewer: ```suggestion
If the recommendations when targeting an Inngest Server [4.2.1] are followed, then presence of the `INNGEST_DEV` environment variable (or equivalent runtime configuration option) MUST turn off signature verification.
```

### 4.2.3. Proxies and routing

The URL through which an SDK communicates with an Inngest Server may be indirect. In these cases, an SDK SHOULD provide the Developer with a method of setting the origin of API requests and Event ingestion.

> We RECOMMEND that a Developer an use the `INNGEST_API_ORIGIN` and `INNGEST_EVENT_API_ORIGIN` environment variables [3.2] for this.

Reviewer: ```suggestion
We RECOMMEND that a Developer can use the `INNGEST_API_ORIGIN` and `INNGEST_EVENT_API_ORIGIN` environment variables [3.2] for this.
```

Guessing

## 4.3. Sync Requests

A Sync is a handshake performed between Inngest and an SDK in which the SDK informs Inngest that a set of Functions belonging to an App can be executed by Inngest at the reported URL.

It is a process that is ideally triggered automatically following every deploy by a Developer, though Inngest may also perform a Sync periodically.

A Sync is always initiated with a `PUT` request, but that initiating request can come from multiple sources, including Inngest.

```mermaid
sequenceDiagram
participant ?
participant SDK
participant Inngest

?->>+SDK: 1. PUT
SDK->>+Inngest: 2. POST<br/>Sync Endpoint
alt 3. Failure
    Inngest-->>SDK: 500
    SDK-->>?: 500
else 4. Success
    Inngest-->>-SDK: 200
    SDK-->>-?: 200
end
```

The numbered statements in the diagram above will be detailed in the following sections.

### 4.3.1. Receiving a Sync Request

> When an SDK receives a `PUT` request, it is treated as the first step of a Sync. This request can come from automatically from Inngest, from being triggered when using an Inngest UI, or from any `PUT` request, allowing a Developer to manually trigger a deploy by contacting their own service.

Reviewer: ```suggestion
When an SDK receives a `PUT` request, it is treated as the first step of a Sync. This request can come automatically from Inngest, from being triggered when using an Inngest UI, or from any `PUT` request, allowing a Developer to manually trigger a deploy by contacting their own service.
```

           * midnight every day.
           */
          cron: string;
        }
    >;

    steps: {
      step: {
        id: "step";
        name: "step";
        runtime: {
          type: "http";

          /**
           * The URL that Inngest should make requests to when wanting to
           * send a Call Request specifically to this Function.
           *
           * It MUST be the `url` defined above, with two query string
           * parameters added:
           *
           *  - `stepId` with the value `step`
           *  - `fnId` with the value of the Function's ID
           */
          url: string;
        };

        retries?: {
          /**
           * The number of times to attempt this step. If this object or
           * property isn't defined, the default is `4` (which is 3 retries).
           */
          attempts?: number;
        };
      };
    };

    /**
     * Allow the specification of an idempotency key using event data. If
     * specified, this overrides the `rateLimit` object.
     */
    idempotency?: string;

    /**
     * batchEvents specifies the batch configuration on when this function

>      * should be invoked when one of the requirements are fulfilled.

Reviewer: ```suggestion
     * should be invoked when one of the requirements is fulfilled.
```

     */
    batchEvents?: {
      /**
       * The maximum number of events to be consumed in one batch,
       * Currently allowed max value is 100.
       */
      maxSize: number;

      /**
       * How long to wait before invoking the function with a list of events.
       * If timeout is reached, the function will be invoked with a batch
       * even if it's not filled up to `maxSize`.
       *
       * MUST be 1s to 60s.
       */
      timeout: string;
    };

    cancel?: Array<{
      /**
       * The name of the event that should cancel the function run.
       */
      event: string;

      /**
       * The expression that must evaluate to true in order to cancel the function run. There
       * are two variables available in this expression:
       * - event, referencing the original function's event trigger
       * - async, referencing the new cancel event.
       */
      if?: string;

      /**
       * An optional timeout that the cancel is valid for.  If this isn't
       * specified, cancellation triggers are valid for up to a year or until the
       * function ends.
       *

>        * The time to wait can be specifed using a Time String or an ISO 8601

Reviewer: ```suggestion
       * The time to wait can be specified using a Time String or an ISO 8601
```

       * date.
       */
      timeout?: string;
    }>;

    /**
     * Rate limit workflows, only running them a given number of times (limit) per
     * period. This can optionally include a `key`, which is used to further
     * constrain throttling similar to idempotency.
     */
    rateLimit?: {
      /**
       * An optional key to use for rate limiting, similar to idempotency.
       */
      key?: string;

      /**
       * The number of times to allow the function to run per the given `period`.
       */
      limit: number;

      /**
       * The period of time to allow the function to run `limit` times.
       */
      period: TimeStr;
    };

    /**
     * Debounce delays functions for the `period` specified. If an event is sent,
     * the function will not run until at least `period` has elapsed.
     *
     * If any new events are received that match the same debounce `key`, the

>      * function is reshceduled for another `period` delay, and the triggering

Reviewer: ```suggestion
     * function is rescheduled for another `period` delay, and the triggering
```

           * - `"ai"`:  references a custom string
           */
          key?: string;

          /**
           * An optional scope for the concurrency group.  By default, concurrency limits are
           * scoped to functions - one function's concurrency limits do not impact other functions.
           *
           * Changing this "scope" allows concurrency limits to work across environments (eg. production
           * vs branch environments) or across your account (global).
           */
          scope?: "fn" | "env" | "account";
        }>;
  }>;
}
```

**Headers**

The SDK MUST adhere to the global header requirements when sending requests [4.1.4].

If an `X-Inngest-Server-Kind` header was set on the sync request received by the SDK, the SDK SHOULD forward the header’s value with a new name of `X-Inngest-Expected-Server-Kind`. This header is used to combat situations where services that a Developer understands to target Inngest Cloud can not be synced with an Inngest Dev Server and vice versa.

### 4.3.3. Handling failure

If Inngest returns a non-`200 OK` status code, the Sync has failed.

> A `400 Bad Request` response indicates that either the App and Function configurations passed were invalid, or the Inngest service reaached was not the one that started the deployment, for example if the Inngest Dev Server starts the deployment, but the SDK reaches out to Inngest Cloud.

Reviewer: ```suggestion
A `400 Bad Request` response indicates that either the App and Function configurations passed were invalid, or the Inngest service reached was not the one that started the deployment, for example if the Inngest Dev Server starts the deployment, but the SDK reaches out to Inngest Cloud.
```

```tsx
{
	error: string;
}
```

Any other status code may not contain the same JSON payload.

An SDK SHOULD pass this error back, closing the `PUT` request, by returning a `500 Internal Server Error` with the following JSON payload:

```tsx
{
	message: "[message_from_response_error]";
	modified: false;
}
```

### 4.3.4. Handling success

If Inngest returns a `200 OK` status code, the Sync has succeeded.

The JSON returned from this request will inform us of whether the Sync added or removed any functions or if this Sync was the same as a previous one. This can be used to log a message for a Developer to see that an App has changed. With this response, the JSON payload from Inngest will be the following:

```tsx
{
	ok: true;
	modified?: boolean;
}
```

An SDK MUST respond to the initial `PUT` request with either the `modified` value passed to it from Inngest, or `false` if none was given. For example:

```tsx
{
	message: "Successfully synced.",
	modified: true;
}
```

## 4.4. Call Requests

A Call Request is a single request from Inngest to an SDK in order to discover work that must be queued or to run blocks of user code defined in Functions or Steps.

> Over the course of an executing a Function that makes no use of Steps, the happy path sees Inngest execute a Function only once, where the Function’s contents are run during that Execution and the result returned to Inngest. If this Function fails by user code throwing an error or some external issue such as network issues or out-of-memory errors, Inngest will retry the Function, by executing it again for each retry.

Reviewer: ```suggestion
Over the course of executing a Function that makes no use of Steps, the happy path sees Inngest execute a Function only once, where the Function’s contents are run during that Execution and the result returned to Inngest. If this Function fails by user code throwing an error or some external issue such as network issues or out-of-memory errors, Inngest will retry the Function, by executing it again for each retry.
```

The API will respond with either a `200 OK` and the JSON body being an object equivalent to the `steps` value in a Call Request body, or a non-`200 OK` response with an `{ error?: string }` JSON object.

If the request was successful, the returned Step data should be used to memoize discovered Steps. If the request failed or returned a non-`200 OK` status code, the SDK MUST respond to the Call Request with a `500 Internal Server Error`.

### 4.4.3. Executing the Function

When an SDK has now validated the request is from Inngest and has retrieved the entire payload, it is ready to run the code represented by the Function.

An SDK SHOULD provide the Developer’s code with select fields from the Call Request’s payload, ensuring that a Developer can utilize these fields to make appropriate decisions in code and change functionality based on the situation. We RECOMMEND that these values are provided, with casing changed to the most idiomatic casing used in a particular SDK:

- `event` - The entire JSON object passed to the Execution.

> - `events` - The entire array os JSON objects passed to the Execution.

Reviewer: ```suggestion
- `events` - The entire array of JSON objects passed to the Execution.
```

- `runId` - Accessed from `ctx.run_id`.
- `attempt` - Accessed from `ctx.attempt`.
- `step` - Step tooling as detailed in Steps [5].

An SDK MUST then call the Function and return either the returned value as JSON or an error to Inngest. When responding, an SDK MUST adhere to the requirements of responding to a request from an Inngest Server [4.1.2].

> A successful response to Inngest MUST always return `200 OK` with a a JSON payload which is the return value of the function.

Reviewer: ```suggestion
A successful response to Inngest MUST always return `200 OK` with a JSON payload which is the return value of the function.
```

}>
```

Each object in the array represents a separate step being reported to the Inngest Server that will be executed in its own Call Request. Note that an SDK MUST also adhere to the requirements when sending or responding to requests [4.1.2].

- `id` is the hashed ID of the Step
- `op` is the string “opcode” (an abbreviation of “operation code”) that specifies the type of Step being reported. The functionality and data expectations of each Step type is different. See Available Step types [5.3].
- `displayName` is an optional string that can be given to provide a human-readable name to the Step in UIs and logging. If this is not specified by the user, it is RECOMMENDED that an SDK set this to the ID for the step that was provided instead.
- `opts` is an object used across many Step types to fine tune their functionality. For information on the expectations of this object for each Step type, see Available Step types [5.3].

### 5.1.2. IDs and hashing

The `id` of each Step MUST be a hex-encoded SHA-1 hash such that the Step can be identified reliably during multiple Call Requests. Commonly, this is done by hashing a human-readable identifier provided by the Developer.

> For example, the human-readable identififer `my-step-id` would be hashed as `e7d8a2f140845095749d60246ff1110c9d01d76a`.

Reviewer: ```suggestion
For example, the human-readable identifier `my-step-id` would be hashed as `e7d8a2f140845095749d60246ff1110c9d01d76a`.
```

## 5.2. Memoizing Step results

The process of memoizing Step results is receiving data or errors that the Inngest Server passes in a Call Request to return that data or error to a Developer’s code, such that they can continue to report more steps or end the function by returning a value or an error.

### 5.2.1. Finding memoized Step data

When a Call Request is made to an SDK, it can contain memoized data from previously-run Steps in the `steps` object passed, which is an object where the keys are hashed Step IDs and the value is either `null` or an object with either a `data` key representing success and the returned data, or an `error` key representing a failed step.

If an object with a `data` key is found, its value will be the result of that particular Step type [5.3], serialized as JSON. An SDK SHOULD make a Developer aware of this serialization process, as first-class entities such as class instances may be lost.

If an object with an `error` key is found, its value will be the final error returned when attempting to run the Step, in the same error format used for reporting function failures:

```tsx
{
	name: string;
	message: string;
	stack?: string;
}
```

> `null` can be also be found depending on the Step type [5.3], where the particular Step type being used indicates its meaning.

Reviewer: ```suggestion
`null` can also be found depending on the Step type [5.3], where the particular Step type being used indicates its meaning.
```

### 5.2.2. Memoizing a Step

As Steps are found by an SDK during a Call Request, it MUST try to find the `data` or `error` already returned for that Step by matching the hashed Step ID with a key in the `steps` object. If this is found, the SDK MUST either return the relevant `data` to the Developer’s call or return/throw/raise the `error`.

If an `error` is returned/thrown/raised and the Developer has not either swallowed the error or returned/thrown/raised a new error, the SDK MUST mark the request as non-retriable [4.4.3].

If a Step’s hash is present in the `steps` object passed in the Call Request, it MUST NOT be reported when responding to that Call Request; only new Steps without a memoized `data` or `error` can be reported.

If a Step is found to have a memoized `error`, an SDK SHOULD create the equivalent of a `StepError` value, passed to the Developer such that they can use it to access the `name`, `message`, and optional `stack` property.

> The `StepError` value SHOULD aim to be used as a language-native error, e.g. JavaScript should extend `Error`, Python should extend `Exception`, and Go should implement the `error` interface. This ensures that the Developer can utilize language features such as `error.As()` r `except StepError as e` to appropriately catch and type the error.

Reviewer: ```suggestion
The `StepError` value SHOULD aim to be used as a language-native error, e.g. JavaScript should extend `Error`, Python should extend `Exception`, and Go should implement the `error` interface. This ensures that the Developer can utilize language features such as `error.As()` or `except StepError as e` to appropriately catch and type the error.
```

If this Step is found, the SDK MUST run the Developer’s code represented by this Step and return either the resulting serialized JSON in the `data.data` key:

```tsx
{
	id: string;
	op: "Step";
	data: {
		// the serialized JSON returned from the Developer's code
		data: any;
	};
	displayName?: string;
}
```

Or the resulting error as serialized JSON:

```tsx
{
	id: string;
	op: "StepError";
	error: {
		name: string;
		message: string;
		stack?: string;
	};
	displayName?: string;
}
```

It is possible that the Step could not be found due to the underlying code changing between reporting the Step and executing it, or a non-deterministic function inserting arbitrary delays between Steps. In either case, if the Step cannot be found within a reasonable period, an SDK MUST return a `"StepNotFound"` operation:

```tsx
{
	id: string;
	op: "StepNotFound";
}
```

The memoized result of the Step will be either a `{ data }` or an `{ error }` object, depending on what was passed by the SDK.

### 5.3.2. Sleep

> A Sleep Step informs the Inngest Server that the Run wishes to be called again after the given amount of time, specified either as an ISO 8601 date as specified [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339), or a “time string.” The latter “time string” is a sequence of a decimal numbers, each with an optional fraction and unit suffix, such as `"300ms"`, or `"2h45m"`. Valid time units are `"ns"`, `"us"` (or `"µs"`), `"ms"`, `"s"`, `"m"`, `"h"`, `"d"`, and `"w"`.

Reviewer: ```suggestion
A Sleep Step informs the Inngest Server that the Run wishes to be called again after the given amount of time, specified either as an ISO 8601 date as specified [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339), or a “time string.” The latter “time string” is a sequence of decimal numbers, each with an optional fraction and unit suffix, such as `"300ms"`, or `"2h45m"`. Valid time units are `"ns"`, `"us"` (or `"µs"`), `"ms"`, `"s"`, `"m"`, `"h"`, `"d"`, and `"w"`.
```
